docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1362 @@
+	// Conditions for the TrainJob
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+```
+
+This table explain rationale for each `TrainJob` parameter:
+
+<table>
+  <tr>
+   <td><strong>Parameter</strong>
+   </td>
+   <td><strong>What is it ?</strong>
+   </td>
+  </tr>
+  <tr>
+   <td><code>TrainingRuntimeRef</code>
+   </td>
+   <td>Reference to the existing <code>TrainingRuntime</code> that pre-deployed by platform engineers
+   </td>
+  </tr>
+  <tr>
+   <td><code>TrainerConfig</code>
+   </td>
+   <td>Configuration for the Trainer such as image, number of nodes, accelerators.
+   </td>
+  </tr>
+  <tr>
+   <td><code>ModelConfig</code>
+   </td>
+   <td>Configuration for the pre-trained model and location for model output
+   </td>
+  </tr>
+  <tr>
+   <td><code>DatasetConfig</code>
+   </td>
+   <td>Configuration for the dataset that will be used to train or fine-tune model
+   </td>
+  </tr>
+  <tr>
+   <td>Labels and Annotations
+   </td>
+   <td>Custom metadata that needs to be applied to the <code>TrainJob</code> resources: JobSet, Job, Pods.
+   </td>
+  </tr>
+  <tr>
+   <td><code>PodSpecOverrides</code>
+   </td>
+   <td>Custom overrides that are specific to the <code>TrainJob</code> and need to be applied to the
+    <code>TrainJob</code> resources. For example, the user identity. Usually, it is managed by
+    custom admission webhooks that inject data to the <code>TrainJob</code> after user creates it
+    via Python SDK or <code>kubectl</code>
+   </td>
+  </tr>
+</table>
+
+### Example of TrainJob
+
+```yaml
+apiVersion: kubeflow.org/v2alpha1
+kind: TrainJob
+metadata:
+  name: torch-ddp
+  namespace: tenant-alpha
+spec:
+  trainingRuntimeRef:
+    name: torch-distributed-multi-node
+  trainerConfig:
+    image: docker.io/custom-training
+    command:
+      - torchrun train.py
+    numNodes: 5
+    resourcesPerNode:
+      requests:
+        nvidia.com/gpu: 2
+```
+
+The above command will be converted as follows:
+
+```bash
+torchrun --nnodes=5 --nproc-per-node=2 train.py
+```
+
+Additionally, the Kubeflow Training SDK allows to create the above `TrainJob` using the Python API:
+
+```python
+def train_func():
+    import torch
+    class Net(torch.nn.Module):
+        """Create the PyTorch Model"""
+        ...
+    model = Net()
+
+    # Attach model to the distributor
+    torch.distributed.init_process_group(backend="nccl")
+    model = torch.nn.parallel.DistributedDataParallel(model)
+
+    # Train model
+    model.train()
+
+# Use Kubeflow SDK to create TrainJob.
+from kubeflow.training import TrainingClient
+
+TrainingClient().train(
+    name="torch-ddp",
+    func=train_func,
+    num_nodes=5,
+    resources_per_node={"gpu": 2},
+)
+```
+
+### Example of LLM Fine-Tuning
+
+This example shows how to create `TrainJob` to fine-tune LLama 7b:
+
+```yaml
+apiVersion: kubeflow.org/v2alpha1
+kind: TrainJob
+metadata:
+  name: tune-llama-with-yelp
+  namespace: tenant-alpha
+spec:
+  trainingRuntimeRef:
```

Comment:
Regarding the model saving output, whether it's S3, Google Storage, or another option, it would be beneficial to have a unified solution that can be utilized by Kubeflow, the Training Operator, Pipelines, or any other Kubeflow components. While this isn't a necessity for V2, it's something we should keep in mind for future development.

Reply:
Agree here, I mentioned something along these lines in one of my comments.

Reply:
+1 to keep flexibility for multiple, pluggable storage proposals

Reply:
> And the path selected by the user will too. For this kind of variable, it can't be a direct part of this training runtime ref (nor can it be hidden so the user doesn't see it and it leads to unexpected behavior).

I think, we should consider that for the fine-tuning user can modify:
- Trainer Hyperparameters (e.g. number of epochs).
- Pre-Trained Model Architecture.
- Dataset location.
- Model location.

As part of my question above, we should discuss how do we want to support LLM runtimes ?
```
What LLM Fine-Tuning Runtimes do we want to add initially. E.g. Llama-7b, Gemma-7b, Llama-70b?
```

For example, if we introduce Llama-70b runtime, user doesn't want to modify the model location. And we also configure set of recipes that are required to fine-tune that model.
I think, we should investigate [the `torchtune` tool](https://github.com/pytorch/torchtune/tree/main), and see if we can re-use it. Since it is natively integrated with `torchrun`.

Additionally, if we introduce more `generic` LLM fine-tuning runtimes, user should be able to configure model, dataset, or trainer with their requirements.

But we are looking for community suggestions on this topic.